text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+---
+layout: default
+title: Fabric Private Chaincode
+nav_order: 3
+---
+ 
+- Feature Name: Fabric Private Chaincode (FPC)
+- Start Date: 2020-01-28)
+- RFC PR: (leave this empty)
+- Fabric Component: Core (fill me in with underlined fabric component, core, orderer/consensus and etc.)
+- Fabric Issue: (leave this empty)
+ 
+# Summary
+[summary]: #summary
+
+This RFC aims to introduce a new security feature for Hyperledger Fabric called Fabric Private Chaincode (FPC) that enhances data confidentiality for chaincodes by executing them in a Trusted Execution Environment (TEE). The first supported TEE platform is Intel SGX, which protects the confidentiality and the integrity of the chaincode execution.
+ 
+FPC is a framework to develop and run chaincodes with strong privacy guarantees.
+With FPC, a chaincode can process transaction arguments and state without revealing/exposing the contents to anybody, including the endorsing peers. That is, clients can encrypt the content of a transaction in a way only the FPC chaincode can decrypt and then process it. The resulting state updates are encrypted by default by the FPC chaincode with keys only known to the chaincode. 
+
+As a result, other parties (both clients and peers) cannot examine the state of the transactions either in the chaincode or on the ledger. Hence, FPC helps to protects sensitive data even from compromised endorsing peers and other unauthorised parties.
+FPC provides the capabilities to verify that a chaincode and the data are protected, that is, Fabric peers and clients can receive cryptographic assurance that the correct chaincode is being run without tampering inside the TEE by means of a hardware-based remote attestation. 
+
+This new model of trust for Fabric smart contracts makes possible high-stakes markets such as private voting systems and sealed-bid auctions; which aren’t supported by the existing model of privacy in Fabric because of the requirement for endorsing peers to be fully trusted for confidentiality.
+ 
+A prototype implementation of FPC is available as Hyperledger Lab on github (https://github.com/hyperledger-labs/fabric-private-chaincode).
+
+# Motivation
```

Comment:
The primary problem solved here is that we protect an organization (and its client) from other organizations (and its peer) and their "nosy-ness".  Fabric does give guarantees for integrity in that scenario but _not_ for confidentiality in the case the organization have to collaborate for a particular chaincode (private data collections do not help in that case). 

Regarding the trusted ledger: it is paramount that a private chaincode works from accurate, consistent and committed ledger data as otherwise you cannot give integrity and hence also no privacy guarantees (see the rollback attack referenced later in the text as one example of why we must provide these guarantees).  As the private chaincode runs in a hostile environment -- local peer is not necessarily trusted -- it cannot rely on the peer doing the validation. While there would be perceivably other ways to get the integrity, e.g., on some consensus among enclaves at other peers in other organizations, our approached seemed by far the most efficient approach.

Reply:
Let me try to understand the problem with an use-case. 

**Use-Case**: Every organization needs to vote on a particular decision by saying either `yes` or `no`. At the end of the voting, no organization should know the vote provided by another organization but only the number `yeses` and `nos`. 

**Smart-Contract**: The smart-contract would receive the vote from the organization and store it on the ledger. Once all organizations in the channel provided their vote, any organization can query the chaincode to count the votes alone. This smart-contract logic is visible to everyone.

**Solution Based on An Existing Approach that Provides Data Confidentiality**: 
     - _implicit collection_: Every organization can put their vote on their individual collection and only hash gets stored on other organization's peers. Note that the vote is passed with some salt so that from the hash it is not possible to determine the vote. In the end, the vote must be revealed to count the number of `yeses` and `nos`. Hence, the implicit collection does not solve the problem.

**Problem**: The above-mentioned use-case and similar use-cases where the organization data must never be revealed outside the organization but using smart-contract and pre-defined logic, another organization can compute something using those secret data. For this scenario, we have no mechanism in Fabric. 
  
**Proposed Solution**: We can use TEE to ensure the confidentiality of the data and still other organization can run some computation on the data. The decryption key is only known to the sgx enclave and shared among all organization's sgx enclave. As a result, only the sgx enclave can decrypt the data, do computation based on the pre-defined logic and respond to the user with the result. In other words, only the CPU chip can see the real data. 

The proposed solution also wants to run the validation component inside the enclave and maintain the ledger data in the enclave. This is to ensure that no space is given for an organization to try any attack and reveal any data. Sure, I will look at the reply attack and get back. 

Let me know if I have understood the problem correctly. If yes, it is good to update the RFC to clearly mention the problem with a use-case. 

Some followup questions: As I asked in some of my previous comments, I would like to understand how the keys are generated? 
- Does it use MRENCVLAE to generate the seal key? 
- Does it use MRSIGNER to generate the seal key and who is the signing authority? 
- How the keys are distributed to other organizations' enclaves?
- How failures are managed?

A lot of question marks around key management and failure handling. It is good to add separate sections to discuss these. I am asking this because based on the key management approach, a certain attack could be possible (Note: my area of expertise is not security but just an enthusiastic).

Reply:
Yes, you got the problem correctly. It is essentially the classical multi-party security problem we are tackling such as millionaires problem and alike.  

Regarding your questions:
- we do seal enclave-local state with key-ing material but externally visible keys (enclave and contract PKs) as well as contract state encryption keys are randomly generated and not via EGETKEY.  
- no, we do not use MRSIGNER but rely only on MRENCLAVE (with MRSIGNER you would have to use some threshold signature scheme to make it work and essentially with the approval process in Fabric v2 lifecycle we get the equivalent)
- enclaves are first registered on the ledger including their attestation on a enclave-specific key.  The first enclave will generate chaincode specific key-material. Via the enclave registry enclaves can then figure out approved enclaves and in mutually authenticated way propagate the keys. 
- regarding failures, this is a bit an open-ended questions. We definitely keep different keys for different chaincodes and using the registry you would know where chaincode would have run and so you can do additional risk management, e.g., by limiting on what chaincodes could be collocated in the same platform.   Not that we do currently would expose actual policies. However, such policies could built on top.

Reply:
I need some more time to digest these points. Failure handling is the most important IMO. In peer, we have a lot of code that is related to recovery after failure and ensuring data consistency. Hence, it would be nice to have a section on failure handling. 

I have not clearly understood how the keys are passed between peers securely, who create key, where it is stored, etc... Maybe once the RFC is updated with some of the figures I have requested to understand these complex concepts, I can re-review the whole proposal again.